loadtest/user/userentity/user.go

```diff
@@ -41,37 +45,67 @@ type Config struct {
 	Password string
 }
 
+type Setup struct {
+	Store     store.MutableUserStore
+	Transport http.RoundTripper
+	Metrics   *perf.UserEntityMetrics
+}
+
 type userTypingMsg struct {
 	channelId string
 	parentId  string
 }
 
+type ueTransport struct {
+	transport http.RoundTripper
+	ue        *UserEntity
+}
+
+func (t *ueTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	startTime := time.Now()
+	resp, err := t.transport.RoundTrip(req)
+	t.ue.observeHTTPRequestTimes(time.Since(startTime).Seconds())
+	if os.IsTimeout(err) {
+		t.ue.incHTTPTimeouts()
+	}
+	if resp != nil && resp.StatusCode >= 400 {
+		t.ue.incHTTPErrors(req.URL.Path, req.Method, resp.StatusCode)
+	}
+	return resp, err
+}
+
 // Store returns the underlying store of the user.
 func (ue *UserEntity) Store() store.UserStore {
 	return ue.store
 }
 
 // New returns a new instance of a UserEntity.
-func New(store store.MutableUserStore, rt http.RoundTripper, config Config) *UserEntity {
-	ue := UserEntity{}
+func New(setup Setup, config Config) *UserEntity {
+	var ue UserEntity
 	ue.config = config
-	ue.client = model.NewAPIv4Client(ue.config.ServerURL)
-	if rt == nil {
-		rt = http.DefaultTransport
+	ue.store = setup.Store
+	ue.metrics = setup.Metrics
+	ue.client = model.NewAPIv4Client(config.ServerURL)
+
+	if setup.Transport == nil {
+		setup.Transport = http.DefaultTransport
 	}
-	ue.client.HttpClient = &http.Client{
-		Transport: rt,
-		Timeout:   5 * time.Second,
```

Comment:
Had to trade this with a lower level timeout (`transport.ResponseHeaderTimeout`) in order to properly catch the error. Overall it doesn't seem to make much of a difference.

Reply:
The *harm* is that unfortunately if the timeout is enforced by the higher layer (`http.Client`) then the error we get from calling `rt.RoundTrip()` will not be typed (a generic `*errors.errorString` is returned) and there would not be a good way to figure out if it's a timeout (unless of course going down the road of using `strings.Contains()` magic which I'd rather avoid).
I've also verified (at the best of my abilities) that `dial` timeout errors are respected so I think we should be fine.

Reply:
Ah I see. Alright then.